woocommerce/woocommerce
# Checkout: Add placeholder option to country and state select, add error validation

After migrating the country and state selection in checkout to a native select recently we uncovered some issues / undefined behaviours that arise due to the different input being used.

This addresses a couple of problems:

1. When the store is set to not choose a default shipping region, we don't want to auto-select the top country option in the select as it often won't make sense, e.g. defaulting to Afghanistan due to alphabetical order will not make sense in most cases. To solve this we introduce an empty placeholder option. It's selected by default in this case but also disabled so it cannot be re-selected

2. Now we have an empty placeholder option it's easy to submit the form without a selected country or state. To solve this, I've introduced client-side validation for country and state when required. I've used existing validation components to keep the styling consistent.


### How to test

0. You'll need a basic store setup with a few in stock products, ensure that this setting is set to "No location by default":

![image](https://github.com/user-attachments/assets/bdae967f-f33f-449d-bbeb-65c54f991d1d)


#### Scenario 1: When "No location by default" is enabled the country select will have a disabled placeholder selected

1. Add a product to cart and proceed from cart to checkout
2. In Checkout see that the country select has a placeholder value selected by default:


<img width="791" alt="Screenshot 2024-07-19 at 3 01 51 PM" src="https://github.com/user-attachments/assets/c4697c2a-dd2e-4de2-bd16-2ab289e706da">
<img width="803" alt="Screenshot 2024-07-19 at 3 01 55 PM" src="https://github.com/user-attachments/assets/04810efe-dff0-4e76-a61d-932e7170db4c">

#### Scenario 2: Submitting "Place order" will validate country is missing

1. From the state of Scenario 1 enter any text in the "City" field. 
2. A validation error will show on the country select:

<img width="799" alt="Screenshot 2024-07-19 at 3 06 23 PM" src="https://github.com/user-attachments/assets/335a9e60-4568-4261-bc93-2a5144576b57">

#### Scenario 3: Selecting a country will remove the validation error

1. From the state of  Scenario 2 click on the country select and select a country, the country validation error should disappear:

<img width="794" alt="Screenshot 2024-07-19 at 3 07 41 PM" src="https://github.com/user-attachments/assets/d72d3d6a-baff-402d-88ef-073c837c1e87">

#### Scenario 4: Select a country with selectable "state" should have an empty option placeholder.

1. From the state of Scenario 3, change the country select to United States
2. The State select should now have a placeholder option in it:

<img width="406" alt="Screenshot 2024-07-19 at 5 33 28 PM" src="https://github.com/user-attachments/assets/dd06db55-d2c7-4e8f-9847-334c4391ed9a">


#### Scenario 5: Trying to submit with a selected country where state is required (but not selected) should show a validation error.

1. With United States selected as in Scenario 4, click "place order"
2. The state select should now show a validation error:

<img width="404" alt="Screenshot 2024-07-19 at 5 34 49 PM" src="https://github.com/user-attachments/assets/3ae861dc-8907-40dc-9c89-b8becb7b6889">


#### Scenario 6: Changing from a country with required state to a country with optional state should not show error

1. With the validation error showing for empty state with United States, change the country selector to Greece
2. The state select should return to a non error state:

<img width="408" alt="Screenshot 2024-07-19 at 5 20 27 PM" src="https://github.com/user-attachments/assets/a29c9320-5def-4c21-9692-a42600c94083">


### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message
Add placeholder options and validation to checkout country and state select inputs.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
